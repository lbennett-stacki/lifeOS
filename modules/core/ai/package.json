{
  "name": "@lifeos/core-ai",
  "version": "0.0.0",
  "private": true,
  "license": "UNLICENSED",
  "main": "./dist/lib.js",
  "types": "./dist/lib.d.ts",
  "type": "module",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc && tsc-alias",
    "clean": "rm -rf dist",
    "lint": "eslint \"src/**/*.ts\" \"tests/**/*.ts\" --fix",
    "dev:tsc": "tsc --watch --preserveWatchOutput",
    "dev:tsc-alias": "tsc-alias --watch",
    "dev": "pnpm build && (concurrently \"pnpm dev:tsc\" \"pnpm dev:tsc-alias\")",
    "format": "prettier --write .",
    "test": "vitest --config vitest.config.ts run"
  },
  "dependencies": {
    "@lifeos/ai": "workspace:../../../packages/ai",
    "@lifeos/config": "workspace:../../../packages/config",
    "@lifeos/di": "workspace:../../../packages/di",
    "@lifeos/logger": "workspace:../../../packages/logger",
    "@lifeos/mailhog": "workspace:../../../tooling/mailhog",
    "dotenv": "^16.4.7",
    "llama-stack-client": "^0.0.35",
    "node-fetch": "^3.3.2",
    "nodemailer": "^6.9.16",
    "openai": "^4.73.1",
    "reflect-metadata": "^0.2.2",
    "zod": "^3.23.8"
  },
  "devDependencies": {
    "@types/node": "^20.17.8",
    "@types/nodemailer": "^6.4.17",
    "@typescript-eslint/eslint-plugin": "^7.18.0",
    "@typescript-eslint/parser": "^7.18.0",
    "concurrently": "^9.1.0",
    "eslint": "^8.57.1",
    "eslint-config-prettier": "^9.1.0",
    "eslint-plugin-prettier": "^5.2.1",
    "prettier": "^3.4.1",
    "tsc-alias": "^1.8.10",
    "typescript": "^5.7.2",
    "vite": "^6.0.5",
    "vite-tsconfig-paths": "^5.1.4",
    "vitest": "^2.1.8"
  }
}
