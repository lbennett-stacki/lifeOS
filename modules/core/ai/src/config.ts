import { z } from 'zod';
import { parseConfig } from '@lifeos/config';
import { logger } from './logger.js';

const smtpSchema = z.object({
  host: z.string(),
  port: z.string(),
  username: z.string(),
  password: z.string(),
});

const configSchema = z.object({
  smtp: smtpSchema,
});

export const config = parseConfig(
  { namespace: '@lifeos/core-ai', schema: configSchema, logger },
  {
    smtp: {
      host: process.env.SMTP_HOST,
      port: process.env.SMTP_PORT,
      username: process.env.SMTP_USERNAME,
      password: process.env.SMTP_PASSWORD,
    },
  },
);
