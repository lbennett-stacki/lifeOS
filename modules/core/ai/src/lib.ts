export * from './agents/lib.js';
export * from './orchestrator.js';
