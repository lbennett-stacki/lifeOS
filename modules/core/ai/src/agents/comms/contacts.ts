import { Agent, AgentCapability, AgentParam } from '@lifeos/ai';
import { z } from 'zod';

@Agent({
  name: 'contacts',
  description: "Access the user's contact list",
})
export class ContactsAgent {
  private readonly contacts = [
    {
      id: '1',
      firstName: 'Lukie',
      middleName: 'Frank',
      lastName: 'Bennetto',
      nicknames: ['The Creator', 'Luke'],
      relationships: ['creator', 'friend', 'foe'],
      email: 'lukie.bennetto@localhost',
      mobile: '+44000000000',
    },
    {
      id: '2',
      firstName: 'Chuckle',
      lastName: 'Buckle',
      nicknames: ['The Brother'],
      relationships: ['brother', 'friend', 'family'],
      email: 'chuckle.buckle@localhost',
      mobile: '+44000000001',
    },
  ];

  @AgentCapability("gets the user's top contacts")
  async topContacts() {
    return this.contacts;
  }

  @AgentCapability("get a user's contact details by id")
  async contactById(
    // TODO: add a cuidSchema
    @AgentParam('a contacts cuid id', z.string())
    id: string,
  ) {
    return this.contacts.find((c) => c.id === id);
  }

  @AgentCapability('fulltext search for a contact')
  async search(
    @AgentParam(
      'a contacts name, nickname or relationship to the user',
      z.string(),
    )
    query: string,
  ) {
    return this.contacts.find((c) => {
      // TODO: perform a real fulltext search
      return (
        c.firstName === query ||
        c.nicknames.includes(query) ||
        c.relationships.includes(query)
      );
    });
  }
}
