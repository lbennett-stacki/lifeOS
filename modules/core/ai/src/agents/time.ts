import { Agent, AgentCapability } from '@lifeos/ai';
import { Inject } from '@lifeos/di';
import { Logger } from '@lifeos/logger';
import { LoggerProviderKey } from '../logger.js';

@Agent({
  name: 'time',
  description: "Access the user's date and time information",
})
export class TimeAgent {
  constructor(@Inject(LoggerProviderKey) private readonly logger: Logger) {}

  @AgentCapability("gets the user's current local time")
  async localDatetime() {
    this.logger.debug('localDatetime test');

    return { datetime: new Date().toLocaleString() };
  }
}
