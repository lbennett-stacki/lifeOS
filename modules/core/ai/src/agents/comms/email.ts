import { config } from '../../config.js';
import { Agent, AgentCapability, AgentParam } from '@lifeos/ai';
import nodemailer from 'nodemailer';
import { z } from 'zod';

@Agent({
  name: 'email',
  description: 'Sends emails',
})
export class EmailAgent {
  constructor(
    private readonly transport = nodemailer.createTransport({
      url: `smtp://${config.smtp.host}:${config.smtp.port}`,
    }),
  ) {}

  @AgentCapability('send an email')
  async sendEmail(
    @AgentParam('the recipients email address', z.string().email())
    to: string,
    @AgentParam('email subject', z.string())
    subject: string,
    @AgentParam('email body', z.string())
    text: string,
  ) {
    console.log('Sending email', {
      to,
      subject,
      text,
    });
    return this.transport.sendMail({
      from: 'lifeos+core-ai@localhost',
      to,
      subject,
      text,
      html: text,
    });
  }
}
