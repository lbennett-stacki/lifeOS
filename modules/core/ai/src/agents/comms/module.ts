import { AgentModule } from '@lifeos/ai';
import { ContactsAgent } from './contacts.js';
import { EmailAgent } from './email.js';

@AgentModule({
  name: 'comms',
  agents: [ContactsAgent, EmailAgent],
})
export class CommsAgentModule {}
