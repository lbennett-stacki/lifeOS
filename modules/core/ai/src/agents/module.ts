import { AgentModule } from '@lifeos/ai';
import { TimeAgent } from './time.js';
import { CommsAgentModule } from './comms/module.js';
import { LoggerProvider } from '../logger.js';

@AgentModule({
  name: 'root',
  providers: [LoggerProvider],
  agents: [TimeAgent],
  modules: [CommsAgentModule],
})
export class RootAgentModule {}
