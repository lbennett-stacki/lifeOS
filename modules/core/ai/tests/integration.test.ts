import 'dotenv/config';
import { expect, test, vi, describe } from 'vitest';
import { AgentApp } from '@lifeos/ai';
import { MailHog } from '@lifeos/mailhog';
import { RootAgentModule } from '../src/agents/module.ts';
import { TimeAgent } from '../src/agents/time.ts';
import { ContactsAgent } from '../src/agents/comms/contacts.ts';
import { EmailAgent } from '../src/agents/comms/email.ts';

const mockDate = new Date(2000, 1, 1, 1, 1, 1, 1);

describe('modules/core/ai integration tests', () => {
  test('DI container resolves agents', async () => {
    vi.useFakeTimers();
    vi.setSystemTime(mockDate);

    const app = new AgentApp(RootAgentModule);

    const timeAgentResolved = app.resolve(TimeAgent);

    expect(timeAgentResolved).toBeDefined();
    expect(await timeAgentResolved.localDatetime()).toMatchSnapshot();

    vi.useRealTimers();
  });

  test('Demo flow A - Send the current time to a contact via email', async () => {
    /*
     * "Send the current time to Luke via email"
     *
     * intent recognition output
     * -------------------------
     * { path: ['time', 'localdatetime'], args: [] },
     * { path: ['comms', 'contacts', 'search'], args: ['Luke'] },
     * { path: ['comms', 'email', 'sendEmail'], args: [email, subject, text] },
     *
     */

    const hog = await MailHog.start();

    const app = new AgentApp(RootAgentModule);

    const timeAgent = app.resolve(TimeAgent);
    const contactsAgent = app.resolve(ContactsAgent);
    const emailAgent = app.resolve(EmailAgent);

    const { datetime: currentTime } = await timeAgent.localDatetime();
    const luke = await contactsAgent.search('Luke');
    if (!luke) {
      throw new Error('Luke not found');
    }
    await emailAgent.sendEmail(
      luke.email,
      `demo flow A`,
      `Hello ${luke.firstName}. The current time is ${currentTime}`,
    );

    expect(luke).toMatchSnapshot();

    await hog.stop();
  });
});
