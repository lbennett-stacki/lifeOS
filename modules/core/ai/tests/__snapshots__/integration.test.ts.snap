// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`modules/core/ai integration tests > DI container resolves agents 1`] = `
{
  "datetime": "2/1/2000, 1:01:01 AM",
}
`;

exports[`modules/core/ai integration tests > Demo flow A - Send the current time to a contact via email 1`] = `
{
  "email": "lukie.bennetto@localhost",
  "firstName": "Lukie",
  "id": "1",
  "lastName": "Bennetto",
  "middleName": "Frank",
  "mobile": "+44000000000",
  "nicknames": [
    "The Creator",
    "Luke",
  ],
  "relationships": [
    "creator",
    "friend",
    "foe",
  ],
}
`;
